sycl/doc/extensions/experimental/sycl_ext_oneapi_memory_export.asciidoc

```diff
@@ -0,0 +1,338 @@
+|Device descriptor |Description
+|`aspect::ext_oneapi_memory_export_linear` | Indicates if the device supports
+the allocation of exportable linear memory and exporting that memory to an
+interoperable handle.
+|======================
+
+=== External Memory Resource Handle Types [[external_mem_res_handles]]
+
+This extension supports two types of external memory resource handles. These
+types are used to import the memory allocated by this extension into external
+APIs.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+
+// External memory resource handle types.
+enum class export_external_mem_handle_type {
+  opaque_fd = 0,
+  win32_nt = 1,
+};
+
+template <export_external_mem_handle_type ExportMemHandleType>
+struct exported_mem;
+
+template <>
+struct exported_mem<export_external_mem_handle_type::opaque_fd> {
+  using type = int;
+};
+
+template <>
+struct exported_mem<export_external_mem_handle_type::win32_nt> {
+  using type = void *;
+};
+
+template <export_external_mem_handle_type ExportMemHandleType>
+using exported_mem_t = typename exported_mem<ExportMemHandleType>::type;
+
+}
+```
+
+The `export_external_mem_handle_type` enum class defines the types of external
+memory resource handles that can be exported by this extension. The `opaque_fd`
+and `win32_nt` values are used during allocation of exportable memory to
+indicate the type of handle that will later be returned by the
+`export_device_mem_handle` function.
+
+The `opaque_fd` handle type corresponds to a POSIX file descriptor, which is
+represented by an `int`.
+
+The `win32_nt` handle type corresponds to a Windows NT handle, which is
+represented by a `void *`.
+
+The `exported_mem` template struct is specialized for each of the
+`export_external_mem_handle_type` values. Each specialization defines a `type`
+alias that represents the type of the exported memory handle for that specific
+handle.
+
+The `exported_mem_t` type alias allows the user to easily obtain the type of
+the exported memory handle based on the `export_external_mem_handle_type`
+template parameter.
+
+=== API of the extension
+
+```c++
+
+namespace sycl::ext::oneapi::experimental {
+
+void *alloc_exportable_device_mem(
+    size_t alignment, size_t size,
+    export_external_mem_handle_type exportMemHandleType,
+    const sycl::device &syclDevice, const sycl::context &syclContext,
+    const property_list& propList = {});
+
+void *alloc_exportable_device_mem(
+    size_t alignment, size_t size,
+    export_external_mem_handle_type exportMemHandleType,
+    const sycl::queue &syclQueue,
+    const property_list& propList = {});
+}
+```
+
+The `alloc_exportable_device_mem` function allocates memory on the device marked
+as having the ability to later export that memory to an external memory resource
+handle.
+
+Memory allocated through this function must only be freed using
+`free_exportable_mem`. Using `sycl::free` to deallocate memory allocated with
+this function results in undefined behavior.
```

Comment:
This doesn't seem very user-friendly.  Today, `sycl::free` can accept any kind of USM pointer, and we've previously acknowledged that it's too difficult for users to track which pointers point to which kinds of allocations.

It would be much nicer to have this work with `sycl::free`.

Reply:
See the comment and linked discussions below on this issue:

https://github.com/intel/llvm/pull/19018#discussion_r2182052526